geo/src/algorithm/concave_hull.rs

```diff
@@ -0,0 +1,432 @@
+use crate::algorithm::convexhull::ConvexHull;
+use crate::algorithm::euclidean_distance::EuclideanDistance;
+use crate::algorithm::euclidean_length::EuclideanLength;
+use crate::prelude::Centroid;
+use crate::utils::partial_min;
+use crate::{Line, LineString, MultiLineString, MultiPoint, MultiPolygon, Point, Polygon};
+use num_traits::Float;
+use rstar::{RTree, RTreeNum};
+use std::collections::VecDeque;
+
+pub trait ConcaveHull<T> {
+    /// Returns a polygon which covers a geometry. Unlike convex hulls, which also cover
+    /// their geometry, a concave hull does so while trying to further minimize its area by
+    /// constructing edges such that the exterior of the polygon incorporates points that would
+    /// be interior points in a convex hull.
+    ///
+    /// This implementation is inspired by https://github.com/mapbox/concaveman
+    /// and also uses ideas from the following paper:
+    /// www.iis.sinica.edu.tw/page/jise/2012/201205_10.pdf
+    ///
+    /// # Example
+    /// ```
+    /// use geo::{line_string, polygon};
+    /// use geo::algorithm::concave_hull::ConcaveHull;
+    ///
+    /// // a square shape
+    /// let poly = polygon![
+    ///     (x: 0.0, y: 0.0),
+    ///     (x: 4.0, y: 0.0),
+    ///     (x: 4.0, y: 4.0),
+    ///     (x: 0.0, y: 4.0),
+    /// ];
+    ///
+    /// // The correct concave hull coordinates
+    /// let correct_hull = line_string![
+    ///     (x: 4.0, y: 0.0),
+    ///     (x: 4.0, y: 4.0),
+    ///     (x: 0.0, y: 4.0),
+    ///     (x: 0.0, y: 0.0),
+    ///     (x: 4.0, y: 0.0),
+    /// ];
+    ///
+    /// let res = poly.concave_hull(2.0);
+    /// assert_eq!(res.exterior(), &correct_hull);
+    /// ```
+    fn concave_hull(&self, concavity: T) -> Polygon<T>
```

Comment:
Given you are definitely going to output a `Polygon` with no interior holes (am I right?), could we make the output as `LineString<T>`?  The user can always make a `Polygon` out of it.  This may just be my personal preference, but we should generally output the simplest type that captures our output.

Reply:
Yeah, you're right about the current implementation avoiding interior holes. I'll switch it to a LineString.